src/tokenserver/auth/oauth.rs

```diff
@@ -0,0 +1,93 @@
+use crate::tokenserver::{error::TokenserverError, settings::Settings};
+
+use core::time::Duration;
+
+/// The information extracted from a valid OAuth token.
+#[derive(Clone, Debug, Default, Deserialize, PartialEq, Serialize)]
+pub struct VerifyOutput {
+    #[serde(rename = "user")]
+    pub fxa_uid: String,
+    pub generation: Option<i64>,
+}
+
+/// The verifier used to verify OAuth tokens.
+#[derive(Clone)]
+pub struct RemoteVerifier {
+    fxa_server_url: String,
+    timeout: u64,
+}
+
+impl RemoteVerifier {
+    const FILENAME: &'static str = "verify.py";
+}
+
+impl From<&Settings> for RemoteVerifier {
+    fn from(settings: &Settings) -> Self {
+        Self {
+            fxa_server_url: settings.fxa_oauth_server_url.clone(),
+            timeout: settings.fxa_oauth_request_timeout,
+        }
+    }
+}
+
+#[async_trait]
+impl VerifyToken for RemoteVerifier {
+    type Output = VerifyOutput;
+
+    /// Verifies an OAuth token. Returns `VerifyOutput` for valid tokens and a `TokenserverError`
+    /// for invalid tokens.
+    async fn verify(&self, token: String) -> Result<VerifyOutput, TokenserverError> {
+        let fxa_server_url = self.fxa_server_url.clone();
+
+        let fut = web::block(move || {
+            let maybe_verify_output_string = Python::with_gil(|py| {
+                let code = include_str!("verify.py");
+                let module = PyModule::from_code(py, code, Self::FILENAME, Self::FILENAME)?;
+                let kwargs = [("server_url", &fxa_server_url)].into_py_dict(py);
+                // `verify.verify_token(token, server_url=fxa_server_url)`
+                let result: &PyAny = module
+                    .getattr("verify_token")?
+                    .call((token,), Some(kwargs))
+                    .map_err(|e| {
+                        e.print_and_set_sys_last_vars(py);
+                        e
+                    })?;
+
+                if result.is_none() {
+                    Ok(None)
+                } else {
+                    let verify_output_python_string = result.downcast::<PyString>()?;
+                    verify_output_python_string.extract::<String>().map(Some)
+                }
+            })
+            .map_err(|_| TokenserverError::invalid_credentials("Unauthorized"))?;
+
+            match maybe_verify_output_string {
+                Some(verify_output_string) => {
+                    serde_json::from_str::<VerifyOutput>(&verify_output_string)
+                        .map_err(|_| TokenserverError::invalid_credentials("Unauthorized"))
+                }
+                None => Err(TokenserverError::invalid_credentials("Unauthorized")),
+            }
+        });
+
+        // The PyFxA OAuth client does not offer a way to set a request timeout, so we set one here
+        // by timing out the future if the verification process blocks this thread for longer
+        // than the specified number of seconds.
+        time::timeout(Duration::new(self.timeout, 0), fut)
```

Comment:
The old Tokenserver [sets a `timeout` attribute on the OAuth client object](https://github.com/mozilla-services/tokenserver/blob/master/tokenserver/verifiers.py#L198), but I don't see such an attribute referenced in the code: https://github.com/mozilla/PyFxA/blob/main/fxa/oauth.py#L24